prog/vnet/rekey_nic.rb

```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+class Prog::Vnet::RekeyNic < Prog::Base
+  subject_is :nic
+
+  def start
+    nic.dst_ipsec_tunnels.each do |tunnel|
+      args = frame["payload"][tunnel.src_nic.id]
+      create_state(tunnel, args)
+    end
+
+    hop :wait_outbound_trigger
+  end
+
+  def wait_outbound_trigger
+    if SemSnap.new(nic.id).set?("trigger_outbound_update")
+      SemSnap.new(nic.id).decr("trigger_outbound_update")
+      hop :setup_outbound
+    end
+    nap 1
+  end
+
+  def setup_outbound
+    nic.src_ipsec_tunnels.each do |tunnel|
+      args = frame["payload"][tunnel.src_nic.id]
+      create_state(tunnel, args)
+      policy_update(tunnel, "out")
+    end
+
+    hop :wait_state_drop_trigger
+  end
+
+  def wait_state_drop_trigger
+    if SemSnap.new(nic.id).set?("old_state_drop")
+      SemSnap.new(nic.id).decr("old_state_drop")
+      hop :old_state_drop
+    end
+
+    nap 1
+  end
+
+  def old_state_drop
+    new_spis = frame["payload"].values.flat_map { |p| [p["spi4"], p["spi6"]] }.flatten
+    state_data = sshable_cmd("sudo ip -n #{nic.src_ipsec_tunnels.first.vm_name(nic)} xfrm state")
+
+    # Extract SPIs along with src and dst from state data
+    states = state_data.scan(/^src (\S+) dst (\S+).*?proto esp spi (0x[0-9a-f]+)/m)
```

Comment:
Does this code turn out worse with `ip --json`, which otherwise should be more stable and robust output, being machine readable in intent/design?

Reply:
--json doesn't work with ip xfrm